File: solutions/haskell/game-of-life/1/src/GameOfLife.hs
module GameOfLife (tick) where

import Data.Maybe (fromMaybe)

{-# INLINEABLE (!?) #-}
(!?) :: [a] -> Int -> Maybe a
xs !? n
  | n < 0 = Nothing
  | otherwise =
      foldr
        ( \x r k -> case k of
            0 -> Just x
            _ -> r (k - 1)
        )
        (const Nothing)
        xs
        n

tick :: [[Int]] -> [[Int]]
tick board =
  [ [cell r c | c <- [0 .. length (board !! r) - 1]]
  | r <- [0 .. length board - 1]
  ]
 where
  cell r c
    | alive && (n == 2 || n == 3) = 1 -- Any live cell with two or three live neighbors lives on.
    | dead && (n == 3) = 1 -- Any dead cell with exactly three live neighbors becomes a live cell.
    | otherwise = 0 -- All other cells die or stay dead.
   where
    alive = ((board !! r) !! c) == 1
    dead = ((board !! r) !! c) == 0
    n = countNeighbours r c

  countNeighbours :: Int -> Int -> Int
  countNeighbours r c =
    sum
      [ el (r + dy) (c + dx)
      | dy <- [-1 .. 1]
      , dx <- [-1 .. 1]
      , (dx, dy) /= (0, 0)
      ]
   where
    el r' c' =
      let row = fromMaybe [] (board !? r')
       in fromMaybe 0 (row !? c')

File: solutions/lua/game-of-life/1/game-of-life.lua
local function get(m, x, y)
	local row = m[x]
	return (row and row[y]) or 0
end

local function neighbors(board, x, y)
	local sum = 0
	for dy = -1, 1 do
		for dx = -1, 1 do
			if not (dx == 0 and dy == 0) then sum = sum + (get(board, x + dx, y + dy)) end
		end
	end
	return sum
end

local function tick(matrix)
	local retval = {}
	for y = 1, #matrix do
		for x = 1, #matrix[y] do
			local n = neighbors(matrix, x, y)
			retval[x] = retval[x] or {}
			if matrix[x][y] == 1 then
				retval[x][y] = (n == 2 or n == 3) and 1 or 0
			else
				retval[x][y] = (n == 3) and 1 or 0
			end
		end
	end
	return retval
end

return { tick = tick }
